packages/datadog-instrumentations/src/couchbase.js

```diff
@@ -0,0 +1,155 @@
+    }))
+
+    emitter.once('error', AsyncResource.bind((error) => {
+      errorChn1qlReq.publish(error)
+      asyncEndChn1qlReq.publish(undefined)
+    }))
+
+    try {
+      return _n1qlReq.apply(this, arguments)
+    } catch (err) {
+      err.stack // trigger getting the stack at the original throwing point
+      errorChn1qlReq.publish(err)
+
+      throw err
+    } finally {
+      endChn1qlReq.publish(undefined)
+    }
+  })
+
+  Bucket.prototype.upsert = wrap('apm:couchbase:upsert', Bucket.prototype.upsert)
+  Bucket.prototype.insert = wrap('apm:couchbase:insert', Bucket.prototype.insert)
+  Bucket.prototype.replace = wrap('apm:couchbase:replace', Bucket.prototype.replace)
+  Bucket.prototype.append = wrap('apm:couchbase:append', Bucket.prototype.append)
+  Bucket.prototype.prepend = wrap('apm:couchbase:prepend', Bucket.prototype.prepend)
+
+  return Bucket
+})
+
+addHook({ name: 'couchbase', file: 'lib/cluster.js', versions: ['^2.6.5'] }, Cluster => {
+  Cluster.prototype._maybeInvoke = wrapMaybeInvoke(Cluster.prototype._maybeInvoke)
+  Cluster.prototype.query = wrapQuery(Cluster.prototype.query)
+
+  shimmer.wrap(Cluster.prototype, 'openBucket', openBucket => function (name, callback) {
+    const ar = new AsyncResource('bound-anonymous-fn')
+    const res = openBucket.apply(this, arguments)
+    bindEmit(res, ar)
+    return res
+  })
+
+  return Cluster
+})
+
+function findCallbackIndex (args) {
+  for (let i = args.length - 1; i >= 2; i--) {
+    if (typeof args[i] === 'function') return i
+  }
+  return -1
+}
+
+function wrapMaybeInvoke (_maybeInvoke) {
+  const wrapped = function (fn, args) {
+    if (!Array.isArray(args)) return _maybeInvoke.apply(this, arguments)
+
+    const callbackIndex = args.length - 1
+    const callback = args[callbackIndex]
+
+    if (callback instanceof Function) {
+      args[callbackIndex] = AsyncResource.bind(callback)
+    }
+
+    return _maybeInvoke.apply(this, arguments)
+  }
+  return shimmer.wrap(_maybeInvoke, wrapped)
+}
+
+function wrapQuery (query) {
+  const wrapped = function (q, params, callback) {
+    const ar = new AsyncResource('bound-anonymous-fn')
+    callback = arguments[arguments.length - 1]
+
+    if (typeof callback === 'function') {
+      arguments[arguments.length - 1] = callback
+    }
+
+    const res = query.apply(this, arguments)
+    bindEmit(res, ar)
+    return res
+  }
+  return shimmer.wrap(query, wrapped)
+}
+
+function wrap (prefix, fn) {
+  const startCh = channel(prefix + ':start')
+  const endCh = channel(prefix + ':end')
+  const asyncEndCh = channel(prefix + ':async-end')
+  const errorCh = channel(prefix + ':error')
+
+  const wrapped = function (key, value, options, callback) {
+    if (
+      !startCh.hasSubscribers
+    ) {
+      return fn.apply(this, arguments)
+    }
+
+    const callbackIndex = findCallbackIndex(arguments)
+
+    if (callbackIndex < 0) return fn.apply(this, arguments)
+
+    const cb = arguments[callbackIndex]
```

Comment:
Is it safe that this is not bound? Any test checking for the correct context?

Reply:
Yes